src/serac/numerics/functional/detail/hexahedron_H1.inl

```diff
@@ -204,85 +228,195 @@ struct finite_element<mfem::Geometry::CUBE, H1<p, c> > {
       gradient(i, 1) = contract<0, 1>(A22, B);
       gradient(i, 2) = contract<0, 1>(A20, G);
     }
+#else
+
+    RAJA_TEAM_SHARED tensor<double, c, q, q, q> value;
+    RAJA_TEAM_SHARED tensor<double, c, dim, q, q, q> gradient;
+    constexpr auto                                   B = calculate_B<apply_weights, q>();
+    constexpr auto                                   G = calculate_G<apply_weights, q>();
+    for (int i = 0; i < c; i++) {
+      RAJA_TEAM_SHARED decltype(deduce_contract_return_type<2, 1>(X[i], B)) A10;
+      RAJA_TEAM_SHARED decltype(deduce_contract_return_type<2, 1>(X[i], G)) A11;
+      RAJA_TEAM_SHARED decltype(deduce_contract_return_type<1, 1>(A10, B))  A20;
+      RAJA_TEAM_SHARED decltype(deduce_contract_return_type<1, 1>(A11, B))  A21;
+      RAJA_TEAM_SHARED decltype(deduce_contract_return_type<1, 1>(A10, G))  A22;
+
+      RAJA::loop<threads_x>(ctx, x_range, [&](int tid) {
+        int qx = tid % BLOCK_X;
+        int qy = (tid / BLOCK_X) % BLOCK_Y;
+        int qz = (tid / (BLOCK_X * BLOCK_Y)) % BLOCK_Z;
+
+        // Perform actual contractions
+        contract<2, 1>(X[i], B, &A10, qx, qy, qz);
+
+        ctx.teamSync();
+        contract<2, 1>(X[i], G, &A11, qx, qy, qz);
+
+        ctx.teamSync();
+        contract<1, 1>(A10, B, &A20, qx, qy, qz);
+
+        ctx.teamSync();
+        contract<1, 1>(A11, B, &A21, qx, qy, qz);
+
+        ctx.teamSync();
+        contract<1, 1>(A10, G, &A22, qx, qy, qz);
+
+        ctx.teamSync();
+
+        contract<0, 1>(A20, B, &value(i), qx, qy, qz);
+
+        ctx.teamSync();
+        contract<0, 1>(A21, B, &gradient(i, 0), qx, qy, qz);
 
+        ctx.teamSync();
+        contract<0, 1>(A22, B, &gradient(i, 1), qx, qy, qz);
+
+        ctx.teamSync();
+        contract<0, 1>(A20, G, &gradient(i, 2), qx, qy, qz);
+
+        ctx.teamSync();
+      });
+    }
+
+#endif
     // transpose the quadrature data into a flat tensor of tuples
+#ifdef USE_CUDA
+    RAJA_TEAM_SHARED
+#endif

```

Comment:
usually, these macros already include the cuda conditional compilation bits, so the user code doesn't have to have write them in their code. Does RAJA not do this for RAJA_TEAM_SHARED?